pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1631 @@
+//
+// All of our base chains have names that are just "${type}-${hook}". e.g., "nat-prerouting".
+type nftablesBaseChain struct {
+	name      string
+	chainType knftables.BaseChainType
+	hook      knftables.BaseChainHook
+	priority  knftables.BaseChainPriority
+}
+
+var nftablesBaseChains = []nftablesBaseChain{
+	// We want our filtering rules to operate on pre-DNAT dest IPs, so our filter
+	// chains have to run before DNAT.
+	{"filter-input", knftables.FilterType, knftables.InputHook, knftables.DNATPriority + "-1"},
+	{"filter-forward", knftables.FilterType, knftables.ForwardHook, knftables.DNATPriority + "-1"},
+	{"filter-output", knftables.FilterType, knftables.OutputHook, knftables.DNATPriority + "-1"},
+	{"nat-prerouting", knftables.NATType, knftables.PreroutingHook, knftables.DNATPriority},
+	{"nat-output", knftables.NATType, knftables.OutputHook, knftables.DNATPriority},
+	{"nat-postrouting", knftables.NATType, knftables.PostroutingHook, knftables.SNATPriority},
+}
+
+// nftablesJumpChains lists our top-level "regular chains" that are jumped to directly
+// from one of the base chains. These are set up from setupNFTables(), and some of them
+// are also referenced in syncProxyRules().
+type nftablesJumpChain struct {
+	dstChain  string
+	srcChain  string
+	extraArgs string
+}
+
+var nftablesJumpChains = []nftablesJumpChain{
+	{kubeEndpointsCheckChain, "filter-input", "ct state new"},
+	{kubeEndpointsCheckChain, "filter-forward", "ct state new"},
+	{kubeEndpointsCheckChain, "filter-output", "ct state new"},
+
+	{kubeForwardChain, "filter-forward", ""},
+
+	{kubeFirewallCheckChain, "filter-input", "ct state new"},
+	{kubeFirewallCheckChain, "filter-output", "ct state new"},
+	{kubeFirewallCheckChain, "filter-forward", "ct state new"},
+
+	{kubeServicesChain, "nat-output", ""},
+	{kubeServicesChain, "nat-prerouting", ""},
+	{kubeMasqueradingChain, "nat-postrouting", ""},
+}
+
+// ensureChain adds commands to tx to ensure that chain exists and doesn't contain
+// anything from before this transaction.
+func ensureChain(chain string, tx *knftables.Transaction, existing sets.Set[string]) {
+	if existing.Has(chain) {
+		return
+	}
+	tx.Add(&knftables.Chain{
+		Name: chain,
+	})
+	tx.Flush(&knftables.Chain{
+		Name: chain,
+	})
+	existing.Insert(chain)
+}
+
+func (proxier *Proxier) setupNFTables(tx *knftables.Transaction) {
+	ipX := "ip"
+	ipvX_addr := "ipv4_addr" //nolint:stylecheck // var name intentionally resembles value
+	noLocalhost := "ip daddr != 127.0.0.0/8"
+	if proxier.ipFamily == v1.IPv6Protocol {
+		ipX = "ip6"
+		ipvX_addr = "ipv6_addr"
+		noLocalhost = "ip6 daddr != ::1"
+	}
+
+	tx.Add(&knftables.Table{
+		Comment: ptr.To("rules for kube-proxy"),
+	})
+
+	// Create and flush base chains
+	for _, bc := range nftablesBaseChains {
+		chain := &knftables.Chain{
+			Name:     bc.name,
+			Type:     ptr.To(bc.chainType),
+			Hook:     ptr.To(bc.hook),
+			Priority: ptr.To(bc.priority),
+		}
+		tx.Add(chain)
+		tx.Flush(chain)
```

Comment:
if the chain already exists , Add() handles the conflict?

Reply:
this is also can use `ensureChain`, no?

Reply:
Yes, following the example of the command-line API, `Create` returns `EEXIST` if the object already exists, but `Add` only returns `EEXIST` if an "incompatible" object already exists (e.g., if there's a chain with the same name but a different `Hook`).

(`Delete` returns `ENOENT` if the object you're deleting doesn't exist. In very new versions of the CLI, there is also a `destroy` command that does delete-or-no-op, but I didn't bother trying to emulate that in knftables because there are a few cases where it's complicated. However, you'll see `tx.Add(x); tx.Delete(x)` in a few places, which has the same effect ("ensure that x does not exist") for the types that it works for.)

Reply:
> this is also can use `ensureChain`, no?

no, that doesn't handle base chains